File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-1.cpp
class Solution {
    vector<pair<int, int>> directions = {{1, 0}, {-1, 0},
                                         {0, 1}, {0, -1}};
public:
    vector<vector<int>> pacificAtlantic(vector<vector<int>>& heights) {
        int ROWS = heights.size();
        int COLS = heights[0].size();
        vector<vector<bool>> pacificOcean(ROWS, vector<bool>(COLS, false));
        vector<vector<bool>> atlanticOcean(ROWS, vector<bool>(COLS, false));


        for (int c = 0; c < COLS; ++c) {
            dfs(0, c, pacificOcean, heights);
            dfs(ROWS - 1, c, atlanticOcean, heights);
        }
        for (int r = 0; r < ROWS; ++r) {
            dfs(r, 0, pacificOcean, heights);
            dfs(r, COLS - 1, atlanticOcean, heights);
        }

        vector<vector<int>> res;
        for (int r = 0; r < ROWS; ++r) {
            for (int c = 0; c < COLS; ++c) {
                if (pacificOcean[r][c] && atlanticOcean[r][c]) {
                    res.push_back({r, c});
                }
            }
        }
        return res;
    }

private:
    void dfs(int r, int c, vector<vector<bool>>& ocean, vector<vector<int>>& heights) {
        ocean[r][c] = true;
        for (auto [dr, dc] : directions) {
            int nr = r + dr, nc = c + dc;
            if (nr >= 0 && nr < heights.size() &&
                nc >= 0 && nc < heights[0].size() &&
                !ocean[nr][nc] && heights[nr][nc] >= heights[r][c]) {
                dfs(nr, nc, ocean, heights);
            }
        }
    }
};

File: Data Structures & Algorithms/longest-increasing-path-in-matrix/submission-1.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        memo = {}
        ROWS, COLS = len(matrix), len(matrix[0])


        def backtrack(r, c):
            if (r, c) in memo:
                return memo[(r, c)]

            res = 1

            directions = [(1,0), (-1,0), (0,1), (0,-1)]

            for dr, dc in directions:
                nr = r + dr
                nc = c + dc

                if 0 <= nr < ROWS and 0 <= nc < COLS:
                    if matrix[nr][nc] > matrix[r][c]:
                        res = max(res, 1 + backtrack(nr, nc))

            memo[(r, c)] = res
            return res
        
        ans = 0
        for r in range(ROWS):
            for c in range(COLS):
                ans = max(ans, backtrack(r, c))

        return ans
